// mp-colorui/dist/h5/components/radio/components/ListRadio.js
import Nerv from "nervjs";
import Taro from "@tarojs/taro-h5";
import { Text, View } from "@tarojs/components";
import { classNames, generateId } from "../../../lib/index";
export default class ListRadio extends Taro.Component {
  render() {
    const props = this.props;

    const [listState, setListState] = Taro.useState(props.list || []);
    Taro.useEffect(() => {
      const list = props.list || [];
      setListState(list.map(item => {
        item.cu_radio_list_id = generateId();
        return item;
      }));
    }, [props.list]);
    const list = listState || [];
    const checkedValue = props.checkedValue;
    const [checked, setChecked] = Taro.useState(checkedValue);
    const listComponent = list.map(item => <View key={item.cu_radio_list_id} className={classNames(["cu-item", {
      checked: checked === item.value,
      disabled: props.disabled
    }])} onClick={() => {
      if (!props.disabled) {
        setChecked(item.value || "");
        props.onChange && props.onChange(item.value);
      }
    }}>
      <View className="content flex align-center">
        <Text>{item.key}</Text>
      </View>
    </View>);
    return <View className={`cu-list menu`}>{listComponent}</View>;
  }

}
ListRadio.defaultProps = {
  list: [],
  checkedValue: "",
  onChange: () => {},
  disabled: false
};
ListRadio.options = {
  addGlobalClass: true
};
